### PR TITLE
Skip FlashMask padding for FA v4 d192/dv128

### PR DESCRIPTION
**Pull request by baoqiwen**
### PR Category

Operator Mechanism

### PR Types

New features

### Description

1. 新增统一的 get_fa_version 函数
2. 根据 fa_version, headdim_q, headdim_v 判断是否需要 pad

### 是否引起精度变化

是

### CODE REVIEW

File: packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py [hunk 1]
```diff
@@ -36,6 +36,71 @@
...
+    if "xpu" in paddle.get_device():
+        return 2
+
+    fa_version = paddle.base.framework.get_flags(["FLAGS_flash_attn_version"])[
+        "FLAGS_flash_attn_version"
+    ]
+
+    deterministic = paddle.get_flags(["FLAGS_cudnn_deterministic"])[
+        "FLAGS_cudnn_deterministic"
+    ]
```
Comment (risemeup1111) [inline packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:82]: P1 **优先级：P1**

这里需要先处理 FA4 backend 是否真的可用。当前文件在 `is_flash_mask_available()` 为 false 时会把 `_flash_attention/_flashmask_attention` 导入为 Paddle fallback，但 `FLAGS_flash_attn_version=4` 且维度满足条件时这里仍返回 4。这样 d192/dv128 会跳过 padding 去调用 fallback；context parallel 还会把 `fa_version=4` 保存到 backward，而 `_flash_mask_available` 为 false 时 backward 没有可执行的 FA4 分支，会直接报错。请在 helper 内把不可用的 FA4 先降级，再继续走 deterministic/维度判断。

```suggestion
    if fa_version == 4 and not is_flash_mask_available():
        fa_version = 3

    deterministic = paddle.get_flags(["FLAGS_cudnn_deterministic"])[
        "FLAGS_cudnn_deterministic"
    ]
```

File: src/paddlefleet/context_parallel_utils.py [hunk 2]
```diff
@@ -720,22 +721,9 @@ def cp_flashmask_allgatherkv_balance_forward(
...
-    ]
-    if fa_version == 3:
-        if "block_mask" in inspect.signature(flashmask_attention).parameters:
-            if deterministic and query.shape[-1] > 128:
-                fa_version = 2
-        elif deterministic:
-            fa_version = 2
+    q_head_dim = query.shape[-1]
+    v_head_dim = value_gathered.shape[-1]
+    fa_version = get_fa_version(q_head_dim, v_head_dim, startend_row_indices)
```
Comment (risemeup1111) [inline src/paddlefleet/context_parallel_utils.py:726]: P1 **优先级：P1**

这里把原来的签名分支全部交给 `get_fa_version()` 后，`flashmask_attention` 不带 `block_mask` 的 Paddle API 变体在 deterministic 模式下不再降级：旧逻辑是 `elif deterministic: fa_version = 2`，仓库里的 `test_branch_c_no_block_mask_deterministic` 也仍然覆盖了 hdim=64 时应返回 2。现在 helper 只在 `head_dim > 128` 时降级，旧 API + deterministic 会返回 3，forward/backward 会调用不支持 deterministic 的 FA3 mask kernel。请保留这个兼容降级，或者同步删除明确不再支持的 API 变体和测试。

```suggestion
    fa_version = get_fa_version(q_head_dim, v_head_dim, startend_row_indices)
    if fa_version == 3:
        sig_params = inspect.signature(flashmask_attention).parameters
        if (
            "block_mask" not in sig_params
            and paddle.get_flags(["FLAGS_cudnn_deterministic"])[
                "FLAGS_cudnn_deterministic"
            ]
        ):
            fa_version = 2
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查当前提交，发现两个需要先处理的兼容性问题，细节已放在行内评论中：一个是 FA4 backend 不可用时仍可能返回 `fa_version=4`，另一个是 context parallel deterministic 模式丢失了旧 API 的降级逻辑。其余检查里 wheel 构建仍在进行中。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py (hunk 1 above)
Comment (risemeup1111) [reply packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:82]: P1 **优先级：P1**

新提交把 refined-recompute 的私有 `_get_fa_version` 删除并改为复用这个 helper（例如 `RefinedRcomputeFlashAttention._first_fwd` 和 `RefinedRcomputeFlashMaskAttention._first_fwd`）。因此这里的 FA4 可用性判断缺失现在也会影响 RR：当 `is_flash_mask_available()` 为 false 但 `FLAGS_flash_attn_version=4` 时，RR 路径会得到 `fa_version == 4`，随后进入 `_flash_attn_fwd/_flash_attn_bwd` 分支，而这些符号在当前文件里不会被导入。请在共享 helper 中统一对不可用 FA4 降级，避免各调用方重复补丁。

File: src/paddlefleet/context_parallel_utils.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/context_parallel_utils.py:726]: P1 **优先级：P1**

新提交同样把 refined-recompute 原有的 `_get_fa_version` 移除了；那个函数也保留了“`flashmask_attention` 签名没有 `block_mask` 且 deterministic 时降级到 FA2”的兼容逻辑。现在 RR masked/非 masked 路径也都依赖共享 `get_fa_version()`，所以如果这里不恢复该签名兼容，旧 Paddle API + deterministic 下不仅 CP 会返回 3，RR 也会走到不支持 deterministic 的 FA3 kernel。建议把这个判断统一放进共享 helper，或在所有调用点显式保留旧逻辑并同步更新相关测试。

File: packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py [hunk 3]
```diff
@@ -73,26 +138,13 @@ def flashmask_attention(
...
+
     need_value_padding = (
-        not (
-            fa_version == 4
-            and use_varlen
-            and is_fa4_support_d192_dv128
-            and q_head_dim == 192
-            and v_head_dim == 128
-        )
+        not (fa_version == 4 and q_head_dim == 192 and v_head_dim == 128)
```
Comment (risemeup1111) [inline packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:147]: P1 **优先级：P1**

这里把 d192/dv128 跳过 padding 的条件放得过宽了。旧逻辑只有在 `fa_version == 4`、调用方显式 `use_varlen=True`，并且底层 `_flashmask_attention` 签名包含 `use_varlen` 时才不 pad；现在只要 `get_fa_version()` 返回 4，`flashmask_attention(..., use_varlen=False)`、普通 `flash_attention()` 以及上层 DPA/RR 路径都会把 `value.shape[-1] == 128` 直接传给原本仍需要 192 维 V 的路径。这样在只支持 varlen/cute d192-dv128 的安装或 fallback 路径上会出现 kernel shape 不匹配或错误输出。请把“是否可以不 pad”作为独立能力判断保留下来，不要仅由 `fa_version` 决定；例如恢复类似下面的条件，并让 `flash_attention`/上层 DPA 只在底层明确支持 head_dim_v 时才跳过 padding：

```python
fa4_d192_dv128_ok = (
    fa_version == 4
    and q_head_dim == 192
    and v_head_dim == 128
    and use_varlen
    and "use_varlen" in inspect.signature(_flashmask_attention).parameters
)
need_value_padding = (not fa4_d192_dv128_ok) and q_head_dim != v_head_dim
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：测试里的旧 `_get_fa_version` 引用基本已迁移，但前两处兼容性问题仍未在生产代码中修复；另外我补充了一条新的行内意见，指出 d192/dv128 跳过 padding 的条件现在放得过宽。具体建议都在行内评论里。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py (hunk 3 above)
Comment (baoqiwen) [reply packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:147]: 最新代码已支持

Comment (risemeup1111) [reply packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:147]: P1 **优先级：P1**

我看了当前 head，线程里的问题还没有被覆盖：`flashmask_attention` 这里仍然只用 `fa_version == 4 and q_head_dim == 192 and v_head_dim == 128` 决定不 pad；同样的条件也出现在普通 `flash_attention` 和上层 DPA 的 padding 分支里。当前代码没有检查调用方是否是 `use_varlen=True`，也没有检查底层 `_flashmask_attention` 是否真的暴露支持 d192/dv128 的能力，因此 non-varlen / 普通 FA / RR 路径仍会直接把 128 维 V 传给可能只接受 192 维 V 的 kernel。

如果底层 FA4 已经全面支持这些路径，请把这个能力显式接到 padding 判断中；否则建议把“是否可以跳过 padding”独立成 helper，并让所有调用点复用，例如：

```python
def _can_skip_d192_dv128_padding(fa_version, q_head_dim, v_head_dim, *, use_varlen=False):
    return (
        fa_version == 4
        and q_head_dim == 192
        and v_head_dim == 128
        and use_varlen
        and "use_varlen" in inspect.signature(_flashmask_attention).parameters
    )
```

File: packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py (hunk 1 above)
Comment (baoqiwen) [reply packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:82]: 我认为，当 FLAGS_flash_attn_version=4 且 fa4 已支持，即进入 FlashMask V4 路径。

若此时 is_flash_mask_available() 返回 False（机器不支持，如 A100），由 FlashMask 内部的断言机制直接拦截报错，不要在外层去改 FA 版本。否则 FA 的版本控制过于混乱。

Comment (risemeup1111) [reply packages/paddlefleet_ops/src/paddlefleet_ops/flash_mask_facade.py:82]: P1 **优先级：P1**

我同意如果用户显式设置 `FLAGS_flash_attn_version=4` 且当前路径必须依赖 FA4，可以选择直接报错；但当前代码还没有把这个语义稳定地落在公共选择点上。现在 `get_fa_version()` 在 `is_flash_mask_available() == False` 时仍会返回 4，而 facade 此时已把 `_flash_attention/_flashmask_attention` 指向 Paddle fallback；RR 文件也只有在 `is_flash_mask_available()` 为 true 时才导入 `_flash_attn_fwd/_flash_attn_bwd`，但后续分支会按 `fa_version == 4` 直接引用这些符号。这样失败形态可能是 fallback 路径拿到未 pad 的 d192/dv128、CP forward/backward 版本不一致，或 RR 中的 `NameError`，而不是一个明确的“当前设备/安装不支持 FA4”的错误。

如果预期是强制 FA4，请在 `get_fa_version()` 或一个所有调用点都会经过的 helper 里前置检查并显式抛错，而不是让各调用点继续收到 4。例如：

```python
if fa_version == 4 and not is_flash_mask_available():
    raise NotImplementedError(
        "FLAGS_flash_attn_version=4 requires the FlashMask FA4 backend; "
        "current device/install does not expose paddlefleet_ops.flash_mask."
    )
```

如果部分路径仍允许降级，则需要在同一个位置返回 3/2，并让测试断言这个行为。关键是不要在 backend 不可用时继续把 `fa_version=4` 传播给 fallback、CP 和 RR。

**Review by umiswing — DISMISSED** [review]

File: src/paddlefleet/refined_recompute/flash_attn.py [hunk 4]
```diff
@@ -21,6 +21,8 @@
 from paddle.autograd import PyLayer
 from paddle.distributed import fleet
 from paddle.nn.functional.flash_attention import flashmask_attention
+from paddlefleet_ops import is_flash_mask_available
+from paddlefleet_ops.flash_mask_facade import get_fa_version
```
Comment (risemeup1111) [inline src/paddlefleet/refined_recompute/flash_attn.py:25]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

当前 head 的 codestyle 已因这次符号迁移失败：`ruff` 报 `src/paddlefleet/refined_recompute/flash_attn.py:1235` 仍调用已删除的 `_get_fa_version`，`1479` 仍引用已删除的 `_flash_mask_available`。这些引用来自当前文件后续的 Ulysses/RR 路径；只在文件顶部改成 `get_fa_version` / `is_flash_mask_available`，但没有同步迁移所有调用，会导致 CI 阻塞，运行时也会 `NameError`。

请把剩余引用一起迁移，形态例如：

```python
_flash_mask_available = is_flash_mask_available()
if _flash_mask_available:
    ...

fa_version = get_fa_version(
    query_states.shape[-1],
    value_states.shape[-1],
    startend_row_indices,
)

assert _flash_mask_available, (...)
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。既有的 P1 线程仍未由新 commit 修复；另外当前 head 的 codestyle 检查出现未定义符号导致失败，我已在行内补充了新的 P1 评论。请先修复这些阻塞项并提交新的 commit，细节见对应行内讨论。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交：上轮提到的 `_get_fa_version` / `_flash_mask_available` 残留引用已修复，codestyle 也已通过。本轮未新增行内评论。

不过既有的 P1 线程仍未由新 commit 修复，当前还不能批准；请继续在对应线程里处理 FA4 backend 可用性、deterministic 兼容回退，以及 d192/dv128 跳过 padding 条件过宽的问题。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dot_product_attention.py [hunk 5]
```diff
@@ -606,18 +618,9 @@ def forward(
...
-                    startend_row_indices_add_sliding_window(
-                        attn_mask_startend_row_indices,
-                        self.sliding_window,
-                        self.head_wise_swa_ratio,
-                        value.shape[2],
-                    )
-                )
-
             if not use_rr_flash_attention and self._has_custom_softmax_scale:
                 extra_kwargs["softmax_scale"] = self.softmax_scale
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dot_product_attention.py:622]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里把自定义 `softmax_scale` 继续透传给 CP FlashMask 时，需要先排除 `fa_version == 2`。当前 DPA 在 `context_parallel_size > 1` 且非 RR 时会把该参数传给 `flashmask_attention_cp`；该 PyLayer 会保存 `ctx.softmax_scale`，backward 再调用 `cp_flashmask_allgatherkv_balance_backward()`。但后者在 `fa_version == 2` 且 `softmax_scale is not None` 时明确抛出 `NotImplementedError`。`get_fa_version()` 在 `FLAGS_flash_attn_version=2`、XPU、或 FA3 deterministic 大 head_dim 等情况下都会返回 2，所以训练会前向通过、反向失败。请在进入 CP FA2 前显式拒绝自定义 scale，或实现 FA2 对该 scale 的完整 forward/backward 支持。

```suggestion
            if (
                self.context_parallel_size > 1
                and fa_version == 2
                and self._has_custom_softmax_scale
            ):
                raise NotImplementedError(
                    "FlashMask context parallel with fa_version==2 does not "
                    "support custom softmax_scale. Disable custom "
                    "softmax_scale or use an FA version that supports it."
                )
            if not use_rr_flash_attention and self._has_custom_softmax_scale:
                extra_kwargs["softmax_scale"] = self.softmax_scale
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查新提交：此前 `_get_fa_version` / `_flash_mask_available` 残留引用导致的 codestyle 阻塞仍保持修复状态；本轮新增的 `softmax_scale` 透传里发现一个新的 CP 反向兼容问题，细节已放在行内评论中。

既有的 FA4 backend 可用性、deterministic 兼容回退、d192/dv128 跳过 padding 条件过宽这三个 P1 线程目前仍未由新 commit 修复，仍需继续处理。当前 wheel 构建失败点是 Docker 镜像拉取超时，暂未作为代码问题评论。

- P3 **优先级：P3** 非行级：PR 标题无法附着到 diff 行。标题仍缺少 PR Category tag，建议改为 `[Operator Mechanism] Skip FlashMask padding for FA v4 d192/dv128`。
  **处理要求：请针对该建议进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by umiswing — APPROVED** [review]

**Review by sneaxiy — APPROVED** [review]
LGTM

**Review by risemeup1 — APPROVED** [review]
